// @denshya/tama/build/index.js
import { EventSignal, Messager, State, Notifier } from '@denshya/reactive';
import { Group } from 'node-group';
import { i as isIterable, a as isAsyncIterable, b as isObservableGetter, c as isJSX, d as isRecord, W as WebNodeBinding, e as isPromiseLike } from './WebNodeBinding-CdoRazHm.js';
import ProtonJSX, { ChildrenType } from './ProtonJSX.js';

const AsyncFunction = async function() {
};
const AsyncGeneratorFunction = async function* () {
};
const AsyncGeneratorPrototype = Object.getPrototypeOf(AsyncGeneratorFunction);

const truthyNonNull = (value) => typeof value === "boolean" ? value : value != null;
function iterableOf(object) {
  if (Symbol.iterator in object) return object;
  const value = object.valueOf();
  if (Symbol.iterator in value) return value;
  throw new TypeError("Given object is not iterable", { cause: { object } });
}
function onDemandRef(factory) {
  let value = null;
  return {
    get current() {
      if (value === null) value = factory();
      return value;
    }
  };
}

class InsertionGroup extends Group {
  constructor() {
    super(...arguments);
    this.placeholder = onDemandRef(() => document.createComment(this.constructor.name));
  }
  replaceChildren(...nodes) {
    if (nodes.length === 0) {
      super.replaceChildren(this.placeholder.current);
    } else {
      super.replaceChildren(...nodes);
    }
  }
  static {
    this.TAG = "insertion-group";
  }
  static {
    if (window.customElements.get(InsertionGroup.TAG) == null) {
      window.customElements.define(InsertionGroup.TAG, InsertionGroup);
    }
  }
}

class MountGuard {
  constructor(element) {
    this.element = element;
    this.properties = /* @__PURE__ */ new Set();
    this.guards = /* @__PURE__ */ new Set();
    this.placeholder = onDemandRef(() => document.createComment(this.constructor.name));
  }
  toggleMount(condition) {
    const placeholder = this.placeholder.current;
    const source = condition ? placeholder : this.element;
    const target = condition ? this.element : placeholder;
    if (source?.parentElement == null) return;
    source.replaceWith(target);
  }
  for(property) {
    if (this.properties.has(property)) return;
    this.properties.add(property);
    property.subscribe((value) => {
      const valid = property.valid?.(value);
      if (valid) {
        this.guards.add(property);
      } else {
        this.guards.delete(property);
      }
      this.toggleMount(this.guards.size >= this.properties.size);
    });
  }
  static is(property) {
    if (property == null) return false;
    if (typeof property !== "object") return false;
    if (typeof property.valid !== "function") return false;
    if (property.subscribe == null) return false;
    return true;
  }
  static truthy(property) {
    if (property.valid(property.get?.() ?? property.valueOf()) === false) {
      return true;
    }
    return false;
  }
}

class Disposal {
  constructor() {
    /** @internal */
    this.controller = onDemandRef(() => new AbortController());
  }
  get signal() {
    return this.controller.current.signal;
  }
  add(effect) {
  }
  adopt(other) {
  }
}

class InflatorAdaptersMap extends Map {
  constructor(inflator, init) {
    super(init);
    this.inflator = inflator;
  }
  add(adapter) {
    if (!super.has(adapter)) {
      const adapterInstance = new adapter(this.inflator);
      super.set(adapter, adapterInstance);
    }
    return this;
  }
}

class Inflator {
  constructor() {
    this.adapters = new InflatorAdaptersMap(this);
  }
  inflate(subject) {
    for (const adapter of this.adapters.values()) {
      adapter.inflator = this;
      const result = adapter.inflate(subject);
      if (result) return result;
    }
    if (subject == null) return subject;
    if (isIterable(subject)) return this.inflateIterable(subject);
    if (isAsyncIterable(subject)) return this.inflateAsyncIterable(subject);
    if (isObservableGetter(subject)) return this.inflateObservable(subject);
    return this.inflatePrimitive(subject);
  }
  /** @internal */
  static cloneWith(inflator, component) {
    const clone = inflator.clone();
    clone.component = component;
    clone.adapters = new InflatorAdaptersMap(clone, inflator.adapters);
    return clone;
  }
}

class TreeContextError extends Error {
}
class TreeContextAPI {
  constructor(parent) {
    this.parent = parent;
    this.contexts = /* @__PURE__ */ new Map();
  }
  provide(context) {
    this.contexts.set(context.constructor, context);
    return context;
  }
  find(context) {
    const instance = this.contexts.get(context);
    if (instance != null) return instance;
    if (this.parent != null) return this.parent.find(context);
    return null;
  }
  /**
   * @throws {TreeContextError}
   */
  require(context) {
    const instance = this.find(context);
    if (instance == null) throw new TreeContextError("No context is provided up the component tree");
    return instance;
  }
  // for<T extends TreeContext>(context: T): Events<{ add(): Observable<InstanceType<T>>, remove(): Observable<void> }> { }
}

class ViewAPI extends EventSignal {
  constructor() {
    super(null);
  }
  async setIterable(iterable) {
    let yieldResult = { done: false, value: void 0 };
    while (yieldResult.done === false) {
      yieldResult = await iterable.next();
      this.set(yieldResult.value);
    }
  }
  /** @internal */
  async initWith(returnResult) {
    if (returnResult == null) return;
    if (returnResult.constructor === AsyncGeneratorPrototype) {
      await this.setIterable(returnResult);
      if (this.current != null) {
        this.default ??= this.current;
      }
      return;
    }
    if (returnResult instanceof Promise) returnResult = await returnResult;
    this.default = returnResult;
    this.set(this.default);
  }
}

class ProtonComponent {
  constructor(inflator, parent) {
    this.inflator = Inflator.cloneWith(inflator, this);
    this.view = new ViewAPI();
    this.tree = new TreeAPI(parent?.tree);
    this.disposal = new Disposal();
  }
}
class TreeAPI {
  constructor(parent) {
    this.parent = parent;
    /** @internal */
    this.thrown = new Messager();
    this.context = new TreeContextAPI(this.parent?.context);
    parent?.thrown.subscribe(this.thrown.dispatch.bind(this.thrown));
  }
  /** @internal */
  caught(thrown) {
    this.thrown.dispatch(thrown);
  }
  catch(callback) {
    void this.thrown.subscribe(callback);
  }
}

class ProtonRef {
  constructor(initialValue) {
    this.current = initialValue;
  }
}
((ProtonRef2) => {
  function resolve(ref, value) {
    if (Array.isArray(ref)) {
      for (const r of ref) resolve(r, value);
      return;
    }
    if (ref instanceof ProtonRef2) {
      ref.current = value;
      return;
    }
    if (ref != null && typeof ref === "object" && "current" in ref) {
      ref.current = value;
      return;
    }
    if (ref instanceof Function) {
      ref(value);
      return;
    }
  }
  ProtonRef2.resolve = resolve;
  function assert(ref) {
    if (ref.current == null) throw new ProtonRefError("Required Ref was not satisfied", ref);
  }
  ProtonRef2.assert = assert;
  function rebase(refs) {
  }
  ProtonRef2.rebase = rebase;
  function from(arg) {
    if (arg instanceof ProtonRef2) return new ProtonRef2(arg.current);
    return new ProtonRef2(arg);
  }
  ProtonRef2.from = from;
})(ProtonRef || (ProtonRef = {}));
class ProtonRefError extends TypeError {
  constructor(message, ref) {
    super(message);
    console.error(ref);
  }
}

const NAMESPACE_SVG = new Set("animate,animateMotion,animateTransform,circle,clipPath,defs,desc,ellipse,feBlend,feColorMatrix,feComponentTransfer,feComposite,feConvolveMatrix,feDiffuseLighting,feDisplacementMap,feDistantLight,feDropShadow,feFlood,feFuncA,feFuncB,feFuncG,feFuncR,feGaussianBlur,feImage,feMerge,feMergeNode,feMorphology,feOffset,fePointLight,feSpecularLighting,feSpotLight,feTile,feTurbulence,filter,foreignObject,g,image,line,linearGradient,marker,mask,metadata,mpath,path,pattern,polygon,polyline,radialGradient,rect,script,set,stop,style,svg,switch,symbol,text,textPath,title,tspan,use,view".split(","));
const NAMESPACE_MATH = new Set("annotation,annotation-xml,maction,math,merror,mfrac,mi,mmultiscripts,mn,mo,mover,mpadded,mphantom,mprescripts,mroot,mrow,ms,mspace,msqrt,mstyle,msub,msubsup,msup,mtable,mtd,mtext,mtr,munder,munderover,semantics".split(","));

var Null;
((Null2) => {
  Null2.ARRAY = Object.freeze(Object.seal([]));
  Null2.OBJECT = Object.freeze(Object.seal({}));
  Null2.FUNCTION = Object.freeze(Object.seal(() => {
  }));
})(Null || (Null = {}));

class WebInflator extends Inflator {
  constructor() {
    super(...arguments);
    this.flags = {
      skipAsync: false,
      disableJSXCache: false
    };
    /**
     * Custom JSX attributes.
     * Adds or Overrides JSX attribute to provide new behavior.
     * These attributes are virtual and won't be presented in the element.
     * */
    this.jsxAttributes = /* @__PURE__ */ new Map();
    this.bindContext = { props: {}, key: "", value: null, bind: (() => {
    }) };
    this.components = [];
  }
  static {
    this.jsxCache = /* @__PURE__ */ new WeakMap();
  }
  clone() {
    const clone = new WebInflator();
    clone.flags = { ...this.flags };
    clone.jsxAttributes = new Map(this.jsxAttributes);
    return clone;
  }
  inflate(subject) {
    if (subject instanceof Node) return subject;
    if (isJSX(subject)) return this.inflateJSXDeeply(subject);
    return super.inflate(subject);
  }
  inflatePrimitive(primitive) {
    return document.createTextNode(primitive);
  }
  inflateFragment() {
    return new Group();
  }
  inflateJSX(jsx) {
    switch (typeof jsx.type) {
      case "string":
        return this.inflateIntrinsic(jsx);
      case "function":
        return this.inflateComponent(jsx.type, jsx.props);
      case "symbol":
        return this.inflateFragment();
    }
    throw new TypeError("Unsupported type of `jsx`", { cause: { jsx } });
  }
  inflateObservable(observable) {
    const value = observable.get?.();
    switch (typeof value) {
      case "object": {
        if (value instanceof Node) return value;
        if (isIterable(value)) return this.inflateIterable(observable);
        if (isJSX(value)) return this.inflateObservableJSX(observable);
        throw new TypeError("Can't choose right way to inflate observable of this type: " + value);
      }
      default:
        return this.inflateObservableText(observable);
    }
  }
  inflateObservableText(observable) {
    const value = observable.get?.();
    const textNode = document.createTextNode(value);
    observable.subscribe((value2) => textNode.nodeValue = observable.get?.() ?? value2);
    return textNode;
  }
  inflateObservableJSX(observable) {
    const placeholder = onDemandRef(() => new Comment("ObservableJSX/" + observable.constructor.name));
    const value = observable.get();
    let element = this.inflateJSXDeeply(value);
    observable.subscribe?.((value2) => {
      const next = this.inflate(value2) ?? placeholder.current;
      element.replaceWith?.(next);
      element = next;
    });
    return element;
  }
  inflateIterable(iterable, parent = new InsertionGroup()) {
    const replace = (otherIterable) => {
      parent.replaceChildren(...this.__inflateIterable__(otherIterable));
    };
    replace(iterableOf(iterable));
    iterable.subscribe?.(replace);
    return parent;
  }
  inflatePositionedIterable(iterable, parent) {
    let anchor = parent.lastChild;
    let currentNodes = [];
    const replace = (otherIterable) => {
      for (const node of currentNodes) node.remove();
      currentNodes.length = 0;
      const nodes = this.__inflateIterable__(otherIterable);
      if (nodes.length === 0 && !anchor) {
        anchor = document.createComment("reactive-iterable");
        parent.appendChild(anchor);
        return;
      }
      if (nodes.length === 0) return;
      currentNodes.push(...nodes);
      if (anchor && anchor.isConnected) {
        anchor.after(...nodes);
      } else {
        parent.prepend(...nodes);
      }
      anchor = nodes[nodes.length - 1];
    };
    replace(iterableOf(iterable));
    iterable.subscribe?.(replace);
  }
  inflateAsyncIterable(asyncIterable) {
    throw new TypeError("Async Iterator is not supported", { cause: { asyncIterable } });
  }
  inflateJSXDeeply(jsx) {
    let inflated;
    if (this.flags.disableJSXCache) {
      inflated = this.inflateJSX(jsx);
    } else {
      const inflatedCached = WebInflator.jsxCache.get(jsx);
      if (inflatedCached != null) return inflatedCached;
      inflated = this.inflateJSX(jsx);
      WebInflator.jsxCache.set(jsx, inflated);
    }
    if (typeof jsx.type === "function") return inflated;
    this.inflateJSXChildren(jsx, inflated);
    return inflated;
  }
  inflateJSXChildren(jsx, parent) {
    const actualParent = parent.nodeType === Node.COMMENT_NODE ? parent.inflated : parent;
    try {
      switch (jsx.childrenType ?? ChildrenType.None) {
        case ChildrenType.None:
          return;
        case ChildrenType.Primitive:
          {
            const text = jsx.props.children;
            if (typeof text !== "object" && typeof text !== "function") {
              actualParent.textContent = text;
            } else {
              WebInflator.subscribeProperty("textContent", text, actualParent);
            }
          }
          return;
        case ChildrenType.ObservableText:
          WebInflator.subscribeProperty("textContent", jsx.props?.children, actualParent);
          return;
        case ChildrenType.ArrayStatic:
          this.inflateIterable(jsx.props.children, actualParent);
          return;
        case ChildrenType.ArrayReactive:
        case ChildrenType.ObservableIterable:
          this.inflatePositionedIterable(jsx.props.children, actualParent);
          return;
        case ChildrenType.VNode:
          actualParent.appendChild(this.inflate(jsx.props.children));
          return;
      }
    } catch (error) {
      console.trace(error, "inflateJSXChildren");
      throw error;
    }
  }
  inflateElement(type, options) {
    if (options?.namespace != null) return document.createElementNS(options.namespace, type, options);
    if (NAMESPACE_SVG.has(type)) return document.createElementNS("http://www.w3.org/2000/svg", type, options);
    if (NAMESPACE_MATH.has(type)) return document.createElementNS("http://www.w3.org/1998/Math/MathML", type, options);
    return document.createElement(type, options);
  }
  /**
   * Creates element and binds properties.
   */
  inflateIntrinsic(jsx) {
    const type = jsx.type;
    const props = jsx.props;
    if (props == null) return this.inflateElement(type);
    const isSVG = NAMESPACE_SVG.has(type);
    const inflated = this.inflateElement(type, props.ns != null ? { namespace: props.ns } : void 0);
    let mountGuard;
    let immediate = false;
    for (const key in props) {
      if (key === "children" || key === "ns") continue;
      const value = props[key];
      if (MountGuard.is(value)) {
        mountGuard ??= new MountGuard(inflated);
        if (MountGuard.truthy(value)) immediate = true;
        mountGuard.for(value);
        continue;
      }
      if (key === "ref") {
        if (value != null) ProtonRef.resolve(value, inflated);
        continue;
      }
      if (key === "className" || key === "class") {
        if (value != null) this.bindClassName(value, inflated, isSVG);
        continue;
      }
      if (key === "style") {
        if (value != null) this.bindStyle(value, inflated.style);
        continue;
      }
      if (key === "on") {
        this.bindEventListeners(value, inflated);
        continue;
      }
      if (key === "aria") {
        if (isRecord(value)) this.bindNodeAria(value, inflated);
        continue;
      }
      if (key === "mounted") {
        if (value != null) {
          value.valid ??= truthyNonNull;
          if (MountGuard.is(value)) {
            if (MountGuard.truthy(value)) immediate = true;
            mountGuard ??= new MountGuard(inflated);
            mountGuard.for(value);
          }
        }
        continue;
      }
      if (typeof value !== "object" && typeof value !== "function" && value != null) {
        if (isSVG || key.includes("-")) inflated.setAttribute(key, value);
        else inflated[key] = value;
      } else if (value != null) {
        if (isSVG || key.includes("-")) WebInflator.subscribeAttribute(inflated, key, value);
        else WebInflator.subscribeProperty(key, value, inflated);
      }
    }
    this.bindFormControls(type, props, inflated);
    this.bindCustomAttributes(props, inflated);
    if (immediate) {
      mountGuard.placeholder.current.inflated = inflated;
      return mountGuard.placeholder.current;
    }
    return inflated;
  }
  bindClassName(cls, node, isSVG) {
    if (typeof cls !== "object") {
      if (isSVG) node.setAttribute("class", String(cls));
      else node.className = String(cls);
    } else if (isSVG) {
      WebInflator.subscribeAttribute(node, "class", cls);
    } else {
      WebInflator.subscribeProperty("className", cls, node);
    }
  }
  bindStyle(style, element) {
    if (isRecord(style)) {
      for (const property in style) {
        if (property.startsWith("--")) {
          WebInflator.subscribe(style[property], function(value) {
            this.setProperty(property, value);
          }, element);
          continue;
        }
        WebInflator.subscribeProperty(property, style[property], element);
      }
      return;
    }
    WebInflator.subscribe(style, function(value) {
      this.cssText = value;
    }, element);
  }
  bindNodeAria(aria, node) {
    for (const key in aria) {
      WebInflator.subscribeProperty(key, aria[key], node);
    }
  }
  bindFormControls(type, props, node) {
    if (type === "input") {
      if (props?.type != null) WebInflator.subscribeProperty("type", props.type, node);
      if (props?.valueAsDate != null) WebNodeBinding.dualSignalBind(node, "valueAsDate", props.valueAsDate, "input");
      if (props?.valueAsNumber != null) WebNodeBinding.dualSignalBind(node, "valueAsNumber", props.valueAsNumber, "input");
      if (props?.value != null) WebNodeBinding.dualSignalBind(node, "value", props.value, "input");
    } else if (type === "textarea") {
      if (props?.value != null) WebNodeBinding.dualSignalBind(node, "value", props.value, "input");
    } else if (type === "select") {
      if (props?.value != null) WebNodeBinding.dualSignalBind(node, "value", props.value, "change");
    }
  }
  bindCustomAttributes(props, node) {
    if (this.jsxAttributes.size === 0) return;
    const bind = (key, value) => {
      WebInflator.subscribeProperty(key, value, node);
    };
    for (const [key, attributeSetup] of this.jsxAttributes) {
      if (key in props === false) continue;
      this.bindContext.bind = bind;
      this.bindContext.props = props;
      this.bindContext.key = key;
      this.bindContext.value = props[key];
      attributeSetup(this.bindContext);
    }
  }
  static {
    this.final = new FinalizationRegistry((disposal) => {
      disposal.controller.current.abort();
    });
  }
  inflateComponent(factory, props) {
    if (this.flags.skipAsync) {
      if (factory instanceof AsyncFunction.constructor) return null;
      if (factory instanceof AsyncGeneratorFunction.constructor) return null;
    }
    if (factory.prototype == null && factory instanceof AsyncFunction.constructor === false) {
      return this.inflate(factory(props));
    }
    const component = new ProtonComponent(this, this.component);
    const componentGroup = new InsertionGroup();
    this.components.push(component);
    WebInflator.final.register(componentGroup, component.disposal);
    try {
      component.view.initWith(factory.call(component, props));
    } catch (thrown) {
      component.tree.caught(thrown);
      console.error(thrown);
      return componentGroup;
    }
    const currentView = component.inflator.inflate(component.view.current);
    replace(currentView);
    function replace(view) {
      if (view == null) componentGroup.replaceChildren();
      if (view instanceof Node) componentGroup.replaceChildren(view);
    }
    let lastAnimationFrame = -1;
    component.view.subscribe((view) => {
      view = component.inflator.inflate(view);
      if (component.view.default == null && component.view.current == null) {
        replace(view);
        return;
      }
      cancelAnimationFrame(lastAnimationFrame);
      lastAnimationFrame = requestAnimationFrame(() => replace(view));
    });
    return componentGroup;
  }
  bindEventListeners(listeners, element) {
    WebInflator.forEachEventBinding(listeners, (event, handler) => {
      if (DELEGATED_EVENTS.has(event)) {
        let ev = element.$EV;
        if (ev == null) {
          ev = {};
          element.$EV = ev;
        }
        let handlers = ev[event];
        if (handlers == null) {
          handlers = [];
          ev[event] = handlers;
        }
        handlers.push(handler);
      } else {
        element.addEventListener(event, handler);
      }
    });
  }
  static forEachEventBinding(source, callback) {
    if (source == null) return;
    if (Array.isArray(source)) {
      for (const entry of source) {
        WebInflator.forEachEventBinding(entry, callback);
      }
      return;
    }
    if (isRecord(source) === false) return;
    for (const key in source) {
      WebInflator.forEachEventBindingValue(key, source[key], callback);
    }
  }
  static forEachEventBindingValue(event, value, callback) {
    if (value == null) return;
    if (Array.isArray(value)) {
      for (const entry of value) {
        WebInflator.forEachEventBindingValue(event, entry, callback);
      }
      return;
    }
    if (WebInflator.isEventListener(value)) {
      callback(event, value);
    }
  }
  static isEventListener(value) {
    if (typeof value === "function") return true;
    if (value instanceof Object && "handleEvent" in value && typeof value.handleEvent === "function") {
      return true;
    }
    return false;
  }
  static {
    this.subscriptions = /* @__PURE__ */ new Map();
  }
  static trackSubscription(target, sub) {
    let subs = WebInflator.subscriptions.get(target);
    if (subs == null) {
      subs = [];
      WebInflator.subscriptions.set(target, subs);
    }
    subs.push(sub);
  }
  /**
   * Binds a property.
   */
  static subscribeProperty(key, source, target) {
    WebInflator.subscribe(source, function(value) {
      this[key] = value;
    }, target);
  }
  /**
   * Binds an attribute.
   */
  static subscribeAttribute(target, key, value) {
    WebInflator.subscribe(value, function(value2) {
      if (value2 != null) {
        this.setAttribute(key, value2);
      } else {
        this.removeAttribute(key);
      }
    }, target);
  }
  /** @internal */
  static subscribe(source, targetBindCallback, thisArg) {
    if (source == null) return;
    if (typeof source !== "object" && typeof source !== "function") {
      targetBindCallback.call(thisArg, source);
      return;
    }
    const unsub = State.subscribeImmediate(source, thisArg != null ? targetBindCallback.bind(thisArg) : targetBindCallback);
    if (unsub != null && thisArg != null) {
      WebInflator.trackSubscription(thisArg, unsub);
    }
    return unsub;
  }
  __inflateIterable__(iterable) {
    if (iterable.length === 0) return Null.ARRAY;
    const result = new Array();
    for (const next of iterable) {
      if (next == null) continue;
      const c = this.inflate(next);
      if (c.nodeType === Node.DOCUMENT_FRAGMENT_NODE && c.childNodes.length === 0) {
        continue;
      }
      result.push(c);
    }
    return result;
  }
}
const DELEGATED_EVENTS = /* @__PURE__ */ new Set([
  "click",
  "dblclick",
  "keydown",
  "keyup",
  "keypress",
  "mousedown",
  "mousemove",
  "mouseup",
  "focusin",
  "focusout",
  "touchstart",
  "touchend",
  "touchmove"
]);
function handleDelegatedEvent(nativeEvent) {
  let target = nativeEvent.target;
  while (target) {
    const ev = target.$EV;
    const handlers = ev?.[nativeEvent.type];
    if (handlers) {
      for (const handler of handlers) {
        handler(nativeEvent);
      }
    }
    if (nativeEvent.cancelBubble) break;
    target = target.parentElement;
  }
}
if (!("$EV" in Node.prototype)) {
  Node.prototype.$EV = void 0;
}
if (!("__tama_delegation" in document)) {
  document.__tama_delegation = true;
  for (const event of DELEGATED_EVENTS) {
    document.addEventListener(event, handleDelegatedEvent);
  }
}

function ProtonDynamic(componentFactory, props) {
  const results = /* @__PURE__ */ new Map();
  const resultsAge = /* @__PURE__ */ new Map();
  function getCachedFactory(component, props2) {
    const propsValues = Object.values(props2).sort();
    const cachedResultKey = results.keys().find((oldPropsValues) => arrayEquals(propsValues, oldPropsValues));
    if (cachedResultKey != null) {
      const cachedResult = results.get(cachedResultKey);
      const keyAge = resultsAge.get(cachedResultKey) ?? 0;
      if (keyAge < Date.now()) {
        results.delete(cachedResultKey);
        resultsAge.delete(cachedResultKey);
      }
      return cachedResult;
    }
    const newResult = component.inflator.inflate(componentFactory(props2));
    results.set(propsValues, newResult);
    resultsAge.set(propsValues, Date.now() + SECONDS_100);
    return newResult;
  }
  function DynamicComponent() {
    const propsState = State.collect(props);
    propsState.subscribe((props2) => this.view.set(getCachedFactory(this, props2)));
    return getCachedFactory(this, propsState.get());
  }
  DynamicComponent.for = componentFactory.name;
  return ProtonJSX.Element(DynamicComponent, null);
}
const SECONDS_100 = 1e6;
function arrayEquals(a, b) {
  return a.length === b.length && a.every((aItem, i) => aItem === b[i]);
}

function ProtonLazy(importFactory) {
  return async () => {
    const module = await importFactory();
    if ("default" in module) return ProtonJSX.Element(module.default, null);
    return ProtonJSX.Element(module, null);
  };
}

class InflatorAdapter {
  constructor(inflator) {
    this.inflator = inflator;
  }
}

class ProtonSwitch {
  constructor(cases) {
    this.cases = cases;
    this.notifier = new Notifier();
    this.current = { key: "default", value: cases["default"] };
  }
  set(key) {
    this.current.key = key;
    this.current.value = this.cases[key];
    this.notifier.dispatch();
  }
  sets(arg) {
    return this.notifier.subscribe(() => {
      if (arg instanceof Function) {
        arg(this.current.value);
      } else {
        arg.set(this.current.value);
      }
    });
  }
  subscribe(next) {
    return this.notifier.subscribe(next);
  }
}
class ProtonSwitchWebInflator extends InflatorAdapter {
  inflate(switcher) {
    if (switcher instanceof ProtonSwitch === false) return;
    let previousKey = switcher.current.key;
    const inflatedCache = {
      [switcher.current.key]: this.inflator.inflate(switcher.current.value)
    };
    switcher.subscribe(() => {
      if (switcher.current.key in inflatedCache === false) {
        inflatedCache[switcher.current.key] = this.inflator.inflate(switcher.current.value);
      }
      const previousInflated = inflatedCache[previousKey];
      const currentInflated = inflatedCache[switcher.current.key];
      if (previousInflated instanceof Element && currentInflated instanceof Element) {
        previousInflated.replaceWith(currentInflated);
      }
      previousKey = switcher.current.key;
    });
    return inflatedCache[switcher.current.key];
  }
}

const Proton = { Lazy: ProtonLazy, Dynamic: ProtonDynamic, Switch: ProtonSwitch, Ref: ProtonRef };

function kebabCase(text) {
  return text.replace(/(\p{Lu}+)/gu, "-$1").replace(/_/g, "-").toLowerCase();
}

class WebJSXSerializer {
  constructor() {
    this.jsxAttributesToString = WebJSXSerializer.jsxAttributesToString;
  }
  /** Inherits customization and options applied to `inflator`. */
  inherit(inflator) {
    inflator.flags.skipAsync = true;
    this.inflator = inflator;
  }
  toString(value) {
    if (value == null) return "";
    if (Array.isArray(value)) return this.iterableToString(value);
    if (typeof Element !== "undefined" && value instanceof Element) return value.outerHTML;
    if (typeof DocumentFragment !== "undefined" && value instanceof DocumentFragment) return this.iterableToString(value.childNodes);
    if (typeof Node !== "undefined" && value instanceof Node) return value.textContent ?? "";
    if (isObservableGetter(value)) return String(value.get());
    if (isIterable(value)) return this.iterableToString(value);
    if (isJSX(value)) {
      if (typeof value.type === "function") {
        return this.componentToString(value.type, value.props);
      }
      return this.jsxToString(value);
    }
    if (value instanceof Object) {
      throw new TypeError("JSX Child can't be object: " + value.constructor);
    }
    return String(value);
  }
  iterableToString(iterable) {
    let children = "", child;
    for (child of iterable) children += this.toString(child);
    return children;
  }
  jsxToString(jsx) {
    if (jsx.props == null) {
      const type2 = jsx.type;
      if (WebJSXSerializer.selfClosingTags[type2]) {
        return "<" + type2 + "/>";
      }
      return "<" + type2 + "></" + type2 + ">";
    }
    if (jsx.props["data-nosnippet"] === true) return "";
    if (jsx.props["data-nosnippet"] === "true") return "";
    const children = this.toString(jsx.props.children);
    if (jsx.type.constructor === Symbol) return children;
    const type = jsx.type;
    const attributes = WebJSXSerializer.jsxAttributesToString.call(this, jsx.props);
    if (WebJSXSerializer.selfClosingTags[type]) {
      return "<" + type + attributes + "/>";
    }
    if (children.length === 0) {
      return "<" + type + attributes + "></" + type + ">";
    }
    return "<" + type + attributes + ">" + children + "</" + type + ">";
  }
  componentToString(factory, props) {
    if (factory instanceof AsyncFunction.constructor) return "";
    if (factory instanceof AsyncGeneratorFunction.constructor) return "";
    return this.toString(factory(props));
  }
}
((WebJSXSerializer2) => {
  function styleToString(style) {
    if (isRecord(style)) {
      let styleString = "";
      for (const propertyName in style) {
        const key = kebabCase(propertyName);
        const value = gettable(style[propertyName]);
        styleString += key + ":" + value + ";";
      }
      return styleString;
    }
    return String(style);
  }
  WebJSXSerializer2.styleToString = styleToString;
  function gettable(value) {
    if (isObservableGetter(value)) return value.get();
    return value;
  }
  WebJSXSerializer2.gettable = gettable;
  function applyCustomJSXAttributes(props) {
    if (this.inflator == null) return;
    if (this.inflator.jsxAttributes.size === 0) return;
    const bind = (key, value) => {
      props[key] = gettable(value);
    };
    for (const key of this.inflator.jsxAttributes.keys()) {
      if (key in props === false) continue;
      const attributeSetup = this.inflator.jsxAttributes.get(key);
      attributeSetup({ props, key, value: props[key], bind });
    }
  }
  WebJSXSerializer2.applyCustomJSXAttributes = applyCustomJSXAttributes;
  function jsxAttributesToString(props) {
    if (props == null) return "";
    applyCustomJSXAttributes.call(this, props);
    let attributes = "", key, value;
    for (key in props) {
      if (key === "on") continue;
      if (key === "ns") continue;
      if (key === "children") continue;
      if (this.inflator?.jsxAttributes.has(key)) continue;
      value = props[key];
      if (value == null) continue;
      if (key === "className") key = "class";
      if (key === "style") value = styleToString(value);
      value = gettable(value);
      if (value == null) continue;
      attributes += " " + key + '="' + value + '"';
    }
    return attributes;
  }
  WebJSXSerializer2.jsxAttributesToString = jsxAttributesToString;
  WebJSXSerializer2.selfClosingTags = {
    "area": true,
    "base": true,
    "br": true,
    "col": true,
    "embed": true,
    "hr": true,
    "img": true,
    "input": true,
    "link": true,
    "meta": true,
    "param": true,
    "source": true,
    "track": true,
    "wbr": true
  };
})(WebJSXSerializer || (WebJSXSerializer = {}));

class WebJSXSerializerAsync {
  /** Inherits customization and options applied to `inflator`. */
  inherit(inflator) {
    this.inflator = inflator;
  }
  async asyncToString(value) {
    if (value == null) return "";
    if (isPromiseLike(value)) return this.asyncToString(await value);
    if (Array.isArray(value)) return this.iterableToStringAsync(value);
    if (typeof Element !== "undefined" && value instanceof Element) return value.outerHTML;
    if (typeof DocumentFragment !== "undefined" && value instanceof DocumentFragment) return this.iterableToStringAsync(value.childNodes);
    if (typeof Node !== "undefined" && value instanceof Node) return value.textContent ?? "";
    if (isObservableGetter(value)) return this.asyncToString(value.get());
    if (isIterable(value)) return this.iterableToStringAsync(value);
    if (isAsyncIterable(value)) return this.asyncIterableToString(value);
    if (isJSX(value)) {
      if (typeof value.type === "function") {
        return this.asyncComponentToString(value.type, value.props);
      }
      return this.asyncJsxToString(value);
    }
    if (value instanceof Object) {
      throw new TypeError("JSX Child can't be object: " + value.constructor);
    }
    return String(value);
  }
  async iterableToStringAsync(iterable) {
    let children = "";
    for (const item of iterable) children += await this.asyncToString(item);
    return children;
  }
  async asyncIterableToString(iterable) {
    return await iterable[Symbol.asyncIterator]().next();
  }
  async asyncComponentToString(factory, props) {
    if (factory instanceof AsyncFunction.constructor) {
      const result = await factory(props);
      return this.asyncToString(result);
    }
    if (factory instanceof AsyncGeneratorFunction.constructor) {
      const iterator = factory(props);
      if (!isAsyncIterable(iterator)) return "";
      const { value } = await iterator.next();
      return await this.asyncToString(value);
    }
    return this.asyncToString(factory(props));
  }
  async asyncJsxToString(jsx) {
    if (jsx.props == null) {
      const type2 = jsx.type;
      if (WebJSXSerializer.selfClosingTags[type2]) return "<" + type2 + "/>";
      return "<" + type2 + "></" + type2 + ">";
    }
    if (jsx.props["data-nosnippet"] === true) return "";
    if (jsx.props["data-nosnippet"] === "true") return "";
    const children = await this.asyncToString(jsx.props.children);
    if (jsx.type?.constructor === Symbol) return children;
    const type = jsx.type;
    const attributes = WebJSXSerializer.jsxAttributesToString.call(this, jsx.props);
    if (WebJSXSerializer.selfClosingTags[type]) return "<" + type + attributes + "/>";
    if (children.length === 0) return "<" + type + attributes + "></" + type + ">";
    return "<" + type + attributes + ">" + children + "</" + type + ">";
  }
}

var Mount;
((Mount2) => {
  const nullable = (value) => !value;
  const nonNullable = (value) => value;
  const patch = (value, valid) => {
    if (typeof value === "object" && value !== null) {
      if ("valid" in value) {
        value = State.from(value);
      }
      value.valid = valid;
    }
    return value;
  };
  Mount2.If = (value, predicate) => patch(value, predicate ?? nonNullable);
  Mount2.Unless = (value) => patch(value, nullable);
})(Mount || (Mount = {}));
const Mount$1 = Mount;

class DOMConnectionObserver {
  // constructor(private readonly methods: DOMConnectionObserverMethods) { }
  constructor(callback) {
    this.callback = callback;
  }
  static {
    this.subscriptions = /* @__PURE__ */ new WeakMap();
  }
  static watch(entries) {
    for (const entry of entries) {
      const subs = this.subscriptions.get(entry.target);
      if (!subs) continue;
      const connected = entry.target.isConnected;
      if (subs.connected === connected) continue;
      subs.connected = connected;
      for (const callback of subs) callback(connected);
    }
  }
  // private static mutationObserver = new MutationObserver(x => this.watch(x))
  // static {
  //   this.mutationObserver.observe(document.documentElement, { childList: true, subtree: true })
  // }
  static observe(node) {
    const subs = /* @__PURE__ */ new Set();
    if (node instanceof Element) {
      DOMConnectionObserver.resizeObserver.observe(node);
      DOMConnectionObserver.intersectionObserver.observe(node);
    }
    return subs;
  }
  static unobserve(node) {
    if (node instanceof Element) {
      this.resizeObserver.unobserve(node);
      this.intersectionObserver.unobserve(node);
    }
  }
  static {
    this.resizeObserver = new ResizeObserver(this.watch.bind(this));
  }
  static {
    this.intersectionObserver = new IntersectionObserver(this.watch.bind(this), { threshold: 0, root: document.body });
  }
  observe(target) {
    const subs = DOMConnectionObserver.subscriptions.getOrInsertComputed(target, DOMConnectionObserver.observe);
    subs.add((isConnected) => this.callback([{ isConnected, target }]));
  }
  unobserve(node) {
    DOMConnectionObserver.subscriptions.delete(node);
    DOMConnectionObserver.unobserve(node);
  }
  // static for(node: Node): Subscriptable<boolean> { }
}

class MountObserver {
  static {
    this.routines = /* @__PURE__ */ new Map();
  }
  static {
    this.observer = new DOMConnectionObserver((entries) => {
      entries.forEach((entry) => {
        const routine = this.routines.get(entry.target);
        if (entry.isConnected === true) routine?.enter();
        if (entry.isConnected === false) routine?.exit();
      });
    });
  }
  static with(routine) {
    return (node) => {
      this.routines.set(node, routine);
      this.observer.observe(node);
      return () => {
        this.routines.delete(node);
        this.observer.unobserve(node);
      };
    };
  }
}

class Lifecycle {
  constructor(arg) {
    this.fsm = {};
    this.abortController = null;
    if (typeof arg === "function") {
      if (arg.length === 1) {
        this.fsm.onEnter = () => {
          this.abortController?.abort();
          this.abortController = new AbortController();
          arg(this.abortController.signal);
        };
        this.fsm.onExit = () => {
          this.abortController?.abort();
          this.abortController = null;
        };
        return;
      }
      this.fsm.onEnter = () => {
        const cleanup = arg();
        this.fsm.onExit = () => {
          cleanup();
          this.fsm.onExit = void 0;
        };
      };
      return;
    }
    this.fsm = { ...arg };
  }
  enter() {
    this.fsm.onEnter?.();
  }
  exit() {
    this.fsm.onExit?.();
  }
}

export { Lifecycle, Mount$1 as Mount, MountObserver, ProtonSwitchWebInflator, ProtonRef as Ref, Proton as Tama, WebInflator, WebJSXSerializer, WebJSXSerializerAsync };
