// @denshya/tama/build/jsx-flat-runtime.js
import ProtonJSX, { ChildrenType } from './ProtonJSX.js';

const INIT = 64;
class VNodePool {
  constructor() {
    this._size = 0;
    this._cap = INIT;
    this._types = new Array(INIT);
    this._props = new Array(INIT);
    this._meta = new Uint8Array(INIT);
    this._parent = new Int32Array(INIT);
    this._firstChild = new Int32Array(INIT);
    this._nextSibling = new Int32Array(INIT);
  }
  allocate(type, rawProps) {
    const i = this._size;
    if (i >= this._cap) this._grow();
    this._types[i] = type;
    this._props[i] = rawProps ?? null;
    const children = rawProps?.children;
    let ct;
    if (children == null) {
      ct = ChildrenType.None;
    } else if (typeof children !== "object" || children === null) {
      ct = typeof children === "function" ? ChildrenType.VNode : ChildrenType.Primitive;
      if (typeof children === "number" && this.isValid(children)) ct = ChildrenType.VNode;
    } else if (children.type != null && !Array.isArray(children)) {
      ct = ChildrenType.VNode;
    } else if (Array.isArray(children)) {
      ct = ChildrenType.ArrayStatic;
      for (const child of children) {
        if (child != null && typeof child === "object" && typeof child.subscribe === "function") {
          ct = ChildrenType.ArrayReactive;
          break;
        }
      }
    } else if (Symbol.iterator in children) {
      ct = ChildrenType.ArrayStatic;
    } else if (typeof children.subscribe === "function") {
      ct = ChildrenType.ObservableText;
    } else {
      ct = ChildrenType.VNode;
    }
    this._meta[i] = ct;
    this._parent[i] = -1;
    this._firstChild[i] = -1;
    this._nextSibling[i] = -1;
    this._size++;
    return i;
  }
  _grow() {
    const cap = this._cap * 2;
    const newMeta = new Uint8Array(cap);
    newMeta.set(this._meta);
    this._meta = newMeta;
    const newParent = new Int32Array(cap);
    newParent.set(this._parent);
    this._parent = newParent;
    const newFirstChild = new Int32Array(cap);
    newFirstChild.set(this._firstChild);
    this._firstChild = newFirstChild;
    const newNextSibling = new Int32Array(cap);
    newNextSibling.set(this._nextSibling);
    this._nextSibling = newNextSibling;
    this._types.length = cap;
    this._props.length = cap;
    this._cap = cap;
  }
  // -- accessors (inlined for perf) --
  getType(i) {
    return this._types[i];
  }
  getProps(i) {
    return this._props[i];
  }
  getChildrenType(i) {
    return this._meta[i];
  }
  getFirstChild(i) {
    return this._firstChild[i];
  }
  getNextSibling(i) {
    return this._nextSibling[i];
  }
  getParent(i) {
    return this._parent[i];
  }
  setParent(i, p) {
    this._parent[i] = p;
  }
  setFirstChild(i, c) {
    this._firstChild[i] = c;
  }
  setNextSibling(i, n) {
    this._nextSibling[i] = n;
  }
  /** Returns true when `i` is a valid, live VNode handle. */
  isValid(i) {
    return Number.isInteger(i) && i >= 0 && i < this._size;
  }
}
const vnodePool = new VNodePool();

const FragmentSymbol = ProtonJSX.FragmentSymbol;
function flatJSX(type, rawProps) {
  const idx = vnodePool.allocate(type, rawProps);
  if (rawProps == null) return idx;
  linkChildren(idx, rawProps.children);
  return idx;
}
function linkChildren(parent, children) {
  if (children == null) return;
  if (typeof children === "number" && vnodePool.isValid(children)) {
    vnodePool.setParent(children, parent);
    vnodePool.setFirstChild(parent, children);
    return;
  }
  if (Array.isArray(children)) {
    let first = -1;
    let prev = -1;
    for (let ci = 0; ci < children.length; ci++) {
      const child = children[ci];
      if (typeof child === "number" && vnodePool.isValid(child)) {
        vnodePool.setParent(child, parent);
        if (prev === -1) {
          first = child;
        } else {
          vnodePool.setNextSibling(prev, child);
        }
        prev = child;
      }
    }
    if (first !== -1) {
      vnodePool.setFirstChild(parent, first);
    }
  }
}

export { ChildrenType, FragmentSymbol as Fragment, flatJSX as jsx, flatJSX as jsxs };
