// @denshya/tama/build/ProtonJSX.js
var ChildrenType = /* @__PURE__ */ ((ChildrenType2) => {
  ChildrenType2[ChildrenType2["None"] = 0] = "None";
  ChildrenType2[ChildrenType2["Primitive"] = 1] = "Primitive";
  ChildrenType2[ChildrenType2["VNode"] = 2] = "VNode";
  ChildrenType2[ChildrenType2["ArrayStatic"] = 3] = "ArrayStatic";
  ChildrenType2[ChildrenType2["ArrayReactive"] = 4] = "ArrayReactive";
  ChildrenType2[ChildrenType2["ObservableText"] = 5] = "ObservableText";
  ChildrenType2[ChildrenType2["ObservableIterable"] = 6] = "ObservableIterable";
  return ChildrenType2;
})(ChildrenType || {});
function classifyChildren(children) {
  if (children == null) return 0 /* None */;
  if (typeof children !== "object" || children === null) {
    switch (typeof children) {
      case "function":
        return 2 /* VNode */;
      default:
        return 1 /* Primitive */;
    }
  }
  if (children instanceof Object && children.type != null && !Array.isArray(children)) {
    return 2 /* VNode */;
  }
  if (Array.isArray(children)) {
    for (const child of children) {
      if (child != null && typeof child === "object" && child.subscribe instanceof Function) {
        return 4 /* ArrayReactive */;
      }
    }
    return 3 /* ArrayStatic */;
  }
  if (children instanceof Object && Symbol.iterator in children) {
    return 3 /* ArrayStatic */;
  }
  if (children.subscribe instanceof Function) {
    return 5 /* ObservableText */;
  }
  return 2 /* VNode */;
}
var ProtonJSX;
((ProtonJSX2) => {
  function Element(type, rawProps) {
    const props = rawProps ?? {};
    const element = { type, props };
    element.childrenType = classifyChildren(props.children);
    return element;
  }
  ProtonJSX2.Element = Element;
  ProtonJSX2.FragmentSymbol = /* @__PURE__ */ Symbol.for("Tama.Fragment");
})(ProtonJSX || (ProtonJSX = {}));
const ProtonJSX$1 = ProtonJSX;

export { ChildrenType, ProtonJSX$1 as default };
