// @denshya/tama/build/WebNodeBinding-CdoRazHm.js
function isRecord(object) {
  return object != null && object.constructor === Object;
}
function isObservableGetter(value) {
  if (value != null && value.subscribe instanceof Function) {
    return true;
  }
  return false;
}
function isIterable(value) {
  return value instanceof Object && value[Symbol.iterator];
}
function isJSX(value) {
  return value != null && !Array.isArray(value) && value.type != null;
}
function isPromiseLike(value) {
  return value != null && typeof value.then === "function";
}
function isAsyncIterable(value) {
  return value != null && typeof value[Symbol.asyncIterator] === "function";
}

const nativeDescriptors = {};
function getNativeDescriptor(instance, property) {
  if (instance instanceof Node === false) {
    throw new TypeError("This type of instance is not supported: " + instance, { cause: { instance } });
  }
  if (instance.constructor.name in nativeDescriptors === false) {
    const descriptor = Object.getOwnPropertyDescriptor(instance.constructor.prototype, property);
    if (descriptor == null) {
      throw new TypeError("This instance constructor does provide a property descriptor for: " + String(property), { cause: { property } });
    }
    nativeDescriptors[instance.constructor.name] = descriptor;
  }
  return nativeDescriptors[instance.constructor.name];
}
var WebNodeBinding;
((WebNodeBinding2) => {
  function dualSignalBind(node, key, value, changeEventKey) {
    const accessor = value;
    if (!isObservableGetter(accessor)) return;
    const descriptor = getNativeDescriptor(node, key);
    if (accessor.get) descriptor.set.call(node, accessor.get());
    if (accessor.set) {
      Object.defineProperty(node, key, {
        configurable: true,
        get: () => descriptor.get.call(node),
        set: (value2) => {
          descriptor.set.call(node, value2);
          accessor.set(value2);
        }
      });
      node.addEventListener(changeEventKey, (event) => accessor.set(event.currentTarget[key]));
    }
    accessor.subscribe?.((value2) => descriptor.set.call(node, accessor.get?.() ?? value2));
  }
  WebNodeBinding2.dualSignalBind = dualSignalBind;
})(WebNodeBinding || (WebNodeBinding = {}));
const WebNodeBinding$1 = WebNodeBinding;

export { WebNodeBinding$1 as W, isAsyncIterable as a, isObservableGetter as b, isJSX as c, isRecord as d, isPromiseLike as e, isIterable as i };
