// node-group/build/index.js
class OrderedList {
  constructor() {
    this.items = [], this.set = /* @__PURE__ */ new Set();
  }
  get length() {
    return this.items.length;
  }
  append(...items) {
    for (const item of items)
      this.set.has(item) || (this.items.push(item), this.set.add(item));
  }
  prepend(...items) {
    for (const item of items)
      this.set.has(item) || (this.items.unshift(item), this.set.add(item));
  }
  replaceItem(oldItem, newItem) {
    if (this.set.has(newItem)) return;
    const index = this.items.indexOf(oldItem);
    index !== -1 && (this.items[index] = newItem, this.set.delete(oldItem), this.set.add(newItem));
  }
  insertBeforeItem(reference, newItem) {
    if (this.set.has(newItem)) return;
    const index = this.items.indexOf(reference);
    index !== -1 && (this.items.splice(index, 0, newItem), this.set.add(newItem));
  }
  getFirst() {
    return this.items[0];
  }
  getLast() {
    return this.items[this.items.length - 1];
  }
  has(item) {
    return this.set.has(item);
  }
  delete(item) {
    const index = this.items.indexOf(item);
    return index === -1 ? !1 : (this.items.splice(index, 1), this.set.delete(item), !0);
  }
  clear() {
    this.items = [], this.set.clear();
  }
  [Symbol.iterator]() {
    return this.items.values();
  }
}
function createNode(value) {
  return typeof value == "string" ? document.createTextNode(value) : value;
}
function NodeListItem(index) {
  return this[index];
}
function CollectionNamedItem(name) {
  return this.find((element) => (element.name || element.id) === name) ?? null;
}
class Group extends HTMLElement {
  constructor() {
    super(...arguments), this.orderedNodes = new OrderedList();
  }
  /** @internal */
  connectedCallback() {
    super.after(...this.orderedNodes), super.remove();
  }
  /** @internal */
  disconnectedCallback() {
  }
  after(...nodes) {
    this.orderedNodes.getLast()?.after?.(...nodes);
  }
  before(...nodes) {
    this.orderedNodes.getFirst()?.before?.(...nodes);
  }
  remove() {
    for (const node of this.orderedNodes)
      node?.remove?.();
  }
  replaceWith(...nodes) {
    this.after(...nodes);
    for (const node of this.orderedNodes)
      node != null && (nodes.includes(node), node.remove?.());
  }
  get children() {
    const collection = [...this.orderedNodes].filter((node) => node instanceof HTMLElement);
    return collection.item = NodeListItem, collection.namedItem = CollectionNamedItem, collection;
  }
  get childNodes() {
    const nodeList = [...this.orderedNodes];
    return nodeList.item = NodeListItem, nodeList;
  }
  appendChild(node) {
    return node === this || (this.orderedNodes.append(node), this.after(node)), node;
  }
  removeChild(child) {
    return this.orderedNodes.delete(child), child;
  }
  replaceChild(node, child) {
    return this.orderedNodes.replaceItem(child, node), child;
  }
  replaceChildren(...nodes) {
    if (nodes.every((node) => this.orderedNodes.has(node))) return;
    const oldNodesSnapshot = [...this.orderedNodes];
    this.append(...nodes);
    for (const oldNode of oldNodesSnapshot)
      nodes.includes(oldNode) || (oldNode.remove?.(), this.orderedNodes.delete(oldNode));
  }
  append(...nodes) {
    const realNodes = nodes.map(createNode);
    this.after(...realNodes), this.orderedNodes.append(...realNodes);
  }
  prepend(...nodes) {
    const realNodes = nodes.map(createNode);
    this.before(...realNodes), this.orderedNodes.prepend(...realNodes);
  }
  insertBefore(node, child) {
    return child != null && this.orderedNodes.insertBeforeItem(child, node), node;
  }
  contains(other) {
    return !!other && this.orderedNodes.has(other);
  }
  get textContent() {
    let result = "";
    for (const node of this.orderedNodes)
      result += node.textContent;
    return result;
  }
  get firstChild() {
    return this.orderedNodes.getFirst() ?? null;
  }
  get lastChild() {
    return this.orderedNodes.getLast() ?? null;
  }
  hasChildNodes() {
    return this.orderedNodes.length > 0;
  }
  static {
    this.TAG = "group-pointer";
  }
  static {
    window.customElements.define(Group.TAG, Group);
  }
}
export {
  Group
};
