// @denshya/reactive/build/index.js
Symbol.subscribe = Symbol.for("subscribe");
function createReactiveSelector(instance) {
  const cache = {};
  return new Proxy(instance, {
    get: (target, key) => {
      const cached = cache[key];
      if (cached != null) return cached;
      let targetValue = target.get();
      target.subscribe((it) => targetValue = it);
      const property = targetValue?.[key];
      if (property instanceof Function) {
        const method = (...args) => property.apply(targetValue, args) === targetValue ? (target.set(targetValue), target) : target.to((value) => property.apply(value, args));
        return cache[key] = method, method;
      }
      const propertyState = target.to((value) => value?.[key]);
      return propertyState.subscribe((propertyValue) => {
        propertyValue !== targetValue?.[key] && (targetValue[key] = propertyValue, target.set(targetValue));
      }), cache[key] = propertyState, propertyState;
    }
  });
}
class ClosureCaptor {
  constructor(closure) {
    this.closure = closure, this.dependencies = /* @__PURE__ */ new Set();
  }
  capture() {
    ClosureCaptor.dispose(this), ClosureCaptor.subscribers.push(this);
    try {
      this.closure();
    } finally {
      ClosureCaptor.subscribers.pop();
    }
  }
  static {
    this.subscribers = [];
  }
  /** Retrieves current (last) subscriber. */
  static getCurrent() {
    return ClosureCaptor.subscribers.at(-1);
  }
  static dispose(subscriber) {
    for (const dependency of subscriber.dependencies)
      dependency.delete(subscriber);
    subscriber.dependencies.clear();
  }
}
class ClosureParticipant {
  constructor() {
    this.dependencies = /* @__PURE__ */ new Set();
  }
  /** Attaches this participant to the current closure captor. */
  attach() {
    const captor = ClosureCaptor.getCurrent();
    captor != null && (this.dependencies.add(captor), captor.dependencies.add(this.dependencies));
  }
  /** Executes previously attached captors. */
  dispatch() {
    for (const subscriber of [...this.dependencies])
      subscriber.capture();
  }
}
var ClosureSignal;
((ClosureSignal2) => {
  class State2 {
    constructor(initialValue) {
      this.closureParticipant = new ClosureParticipant(), this.value = initialValue;
    }
    get current() {
      return this.value;
    }
    get() {
      return this.closureParticipant.attach(), this.value;
    }
    set(nextValue) {
      this.value = nextValue, this.closureParticipant.dispatch();
    }
  }
  ClosureSignal2.State = State2;
  class Computed {
    constructor(factoryClosure) {
      this.state = new State2(null);
      const closure = () => this.state.set(factoryClosure());
      new ClosureCaptor(closure).capture();
    }
    // @ts-expect-error private `value`.
    get() {
      return this.state.value;
    }
    // @ts-expect-error private `value`.
    get current() {
      return this.value;
    }
  }
  ClosureSignal2.Computed = Computed;
})(ClosureSignal || (ClosureSignal = {}));
const finalization = new FinalizationRegistry((unsubscribe) => unsubscribe());
class Messager {
  constructor() {
    this.callbacks = /* @__PURE__ */ new Set();
  }
  dispatch(value) {
    this.callbacks.forEach((callback) => callback(value));
  }
  subscribe(next, options) {
    this.callbacks.add(next);
    const unsubscribe = () => void this.callbacks.delete(next);
    return finalization.register(this, unsubscribe), options?.signal?.addEventListener("abort", unsubscribe), { unsubscribe };
  }
  toJSON() {
  }
}
class Notifier extends Messager {
}
function isObservableGetter(value) {
  return value instanceof Object && "get" in value && (Symbol.subscribe in value || "subscribe" in value);
}
function isObservableLike(value) {
  return value instanceof Object && (Symbol.subscribe in value || "subscribe" in value);
}
class Signal {
  constructor(initialValue) {
    this.closureParticipant = new ClosureParticipant(), this.messager = new Messager(), this.value = initialValue;
  }
  /** Retrieves current `value` without side-effects. Alias to `get`. */
  get current() {
    return this.value;
  }
  /** 
   * Besides returning `value`, sends itself to the capturing closure signal
   * to act as if `Signal` is directly subscribed.
   * 
   * @example
   * const state1 = new Signal(123)
   * const state2 = new Signal(456)
   * const state3 = Signal.capture(() => state1.use() + state2.use())
   * */
  use() {
    return this.closureParticipant.attach(), this.value;
  }
  /** Retrieves current `value` without side-effects. */
  get() {
    return this.value;
  }
  /** Sets given `newValue` or transforms current version of `value`, propagates updates to subscribers. */
  set(newValue) {
    const value = newValue instanceof Function ? newValue(this.value) : newValue;
    this.value = value, this.messager.dispatch(value), this.closureParticipant.dispatch();
  }
  /** Invokes `callback` when `observable` updates. */
  subscribe(callback, options) {
    return this.messager.subscribe(callback, options);
  }
  [Symbol.subscribe](next, options) {
    return this.messager.subscribe(next, options);
  }
  toJSON() {
    return this.value;
  }
  valueOf() {
    return this.value;
  }
}
((Signal2) => {
  function subscribe(value, callback, options) {
    let unsubscribe = { unsubscribe: () => {
    } };
    return value == null || !(value instanceof Object) || isObservableLike(value) === !1 || ("subscribe" in value ? unsubscribe = value.subscribe(callback) : Symbol.subscribe in value && (unsubscribe = value[Symbol.subscribe](callback)), options?.signal?.addEventListener("abort", unsubscribe.unsubscribe)), unsubscribe;
  }
  Signal2.subscribe = subscribe;
  function use(value) {
    if (value instanceof Signal2) return value.use();
    if (isObservableGetter(value)) {
      const anonymousParticipant = new ClosureParticipant();
      return anonymousParticipant.attach(), "subscribe" in value ? value.subscribe(() => anonymousParticipant.dispatch()) : Symbol.subscribe in value && value[Symbol.subscribe](() => anonymousParticipant.dispatch()), value.get();
    }
    return value;
  }
  Signal2.use = use;
  function get(value) {
    if (value instanceof Function) return value();
    if (value instanceof Object) {
      if ("get" in value) return value.get();
      if ("current" in value) return value.current;
    }
    return value;
  }
  Signal2.get = get;
})(Signal || (Signal = {}));
class State extends Signal {
  /** Subscribes to updates and immediately calls the `callback`. */
  subscribeImmediate(callback, options) {
    return options?.signal?.aborted || callback(this.value), this.subscribe(callback, options);
  }
  /**
   * Sets value of this state to `other` one continuously.
   * 
   * @example
   * ```ts
   * const state1 = new State(123)
   * const state2 = new State(0)
   * state1.sets(state2)
   * state2 // Becomes `123`.
   * ```
   * 
   * @example
   * Can be combined with `from` method
   * ```ts
   * const state1 = new State("123")
   * const state2 = new State(0)
   * state1.sets(state2.from(parseInt))
   * ```
   */
  sets(other) {
    return this.messager.subscribe((value) => other.set(value));
  }
  /** Copies value from `other` Signal-like structure just once. */
  copy(other) {
    if ("current" in other) {
      this.set(other.current);
      return;
    }
    this.set(other.get());
  }
  /**
   * @example
   * ```ts
   * const app = new State({ user: { name: "test" } })
   * // Observable Access
   * app.$.user.$.name instanceof State // true
   * // Usage of `$` is cached and only created once.
   * app.$.user === app.$.user // true
   * // Setting to accessed property actually changes it.
   * app.$.user.$.name.set("me")
   * app.get().user.name === "me" // true
   * ```
   */
  get $() {
    return this._$ == null && (this._$ = createReactiveSelector(this)), this._$;
  }
  /**
   * `to` is an alternative name of `map` function because `state.map(x => x.map(e => e))` looks weird.
   * But looks neat as Tacit programming - `string.to(Number)`
   * 
   * Creates a new instance, transforms the value and assigns it.
   * 
   * 
   * @example
   * Useful when to select a value with reactivity:
   * ```ts
   * const ypx = new State("15px")
   * const y = ypx.to(parseFloat)
   * 
   * ypx.set("16px") // Will set `y` to `16`.
   * y.set(15) // Will not affect `ypx`.
   *   ```
   */
  to(predicate) {
    const fork = new State(predicate(this.value));
    return this.messager.subscribe((value) => {
      const newValue = predicate(value);
      newValue !== fork.value && fork.set(newValue);
    }), fork;
  }
  /**
   * Useful when you want to fit a "source" ("sink") to the state.
   * By providing `predicate` function, the source value can transformed and set automatically.
   * 
   * @example
   * ```ts
   * const pointerX = new State(0)
   * 
   * window.addEventListener("pointermove", pointerX.from(event => event.x)) // Or
   * window.when("pointermove").subscribe(pointerX.from(event => event.x))
   * 
   * const event = new State(new PointerEvent(""))
   * 
   * event.sets(pointerX.from(event => event.x)) // Or
   * event.subscribe(pointerX.from(event => event.x))
   * ```
   * 
   * This literally says "`event` sets `value` from `event.currentTarget.value`".
  */
  from(predicate) {
    const sourcing = (value) => this.set(predicate(value));
    return sourcing.set = sourcing, sourcing.next = sourcing, sourcing;
  }
  is(arg1) {
    return arg1 instanceof Function ? { get: () => arg1(this.value), [Symbol.subscribe]: (next) => this.messager.subscribe((value) => next(arg1(value))) } : State.combine([this, arg1], (value, arg12) => value === arg12).readonly;
  }
  get readonly() {
    return { get: this.get.bind(this), [Symbol.subscribe]: this[Symbol.subscribe].bind(this) };
  }
  get writeonly() {
    return { set: this.set.bind(this) };
  }
  /** Captures next state value. */
  get upcoming() {
    let subscription;
    return new Promise((resolve) => {
      subscription = this.subscribe((value) => {
        resolve(value), subscription.unsubscribe();
      });
    });
  }
}
((State2) => {
  function subscribeImmediate(value, callback, options) {
    const result = State2.subscribe(value, callback, options);
    if (options?.signal?.aborted) return result;
    const unwrappedValue = State2.get(value);
    return unwrappedValue === value && unwrappedValue instanceof Object && isObservableLike(unwrappedValue) || callback(unwrappedValue), result;
  }
  State2.subscribeImmediate = subscribeImmediate;
  function capture(closure) {
    const signal = new State2(null), signalClosure = () => signal.set(closure());
    return new ClosureCaptor(signalClosure).capture(), signal;
  }
  State2.capture = capture;
  function combine(states, predicate) {
    const values = states.map(State2.get), computed = new State2(predicate(...values));
    return states.forEach((state, index) => {
      State2.subscribe(state, (value) => {
        values[index] = value, computed.set(predicate(...values));
      });
    }), computed;
  }
  State2.combine = combine;
  function collect(arg1) {
    if (arg1 instanceof Array)
      return State2.combine(arg1, (...values) => values);
    const result = {}, record = new State2(result);
    for (const key in arg1) {
      const value = arg1[key];
      if (isObservableGetter(value) === !1) {
        result[key] = value;
        continue;
      }
      result[key] = value.get(), State2.subscribe(value, (it) => {
        result[key] = it, record.set(result);
      });
    }
    return record;
  }
  State2.collect = collect;
  function f(strings, ...values) {
    return State2.combine(values, (...values2) => strings.map((string, i) => string + String(values2[i] ?? "")).join(""));
  }
  State2.f = f;
  function from(item) {
    if (item instanceof State2) return new State2(item.get());
    if (isObservableGetter(item)) {
      const state = new State2(item.get());
      return "subscribe" in item ? item.subscribe((value) => state.set(value)) : Symbol.subscribe in item && item[Symbol.subscribe]((value) => state.set(value)), state;
    }
    if (isObservableLike(item)) {
      const state = new State2(void 0);
      return State2.subscribe(item, (value) => state.set(value)), state;
    }
    return new State2(State2.get(item));
  }
  State2.from = from;
  async function* asyncIterableOf(state) {
    for (yield await state.get(); ; ) yield await state.upcoming;
  }
  State2.asyncIterableOf = asyncIterableOf;
})(State || (State = {}));
class StateArray extends Signal {
  constructor(init = []) {
    super(Signal.get(init)), Signal.subscribe(init, (value) => this.set(value));
  }
  at(index, mode) {
    return mode === "direct" ? State.combine([this, index], (array, index2) => array[index2]) : State.combine([this, index], (array, index2) => array.at(index2));
  }
  /** Adds value to the end of the array and notifies subscribers. */
  push(value) {
    const index = this.value.push(value);
    return this.messager.dispatch(this.value), index;
  }
  /** Behaves the same way that `Array` does. Transforms every value with `predicate` and create new instance of `StateArray`. */
  map(predicate) {
    const mapped = new StateArray(this.value.map(predicate));
    return this.subscribe((value) => mapped.set(value.map(predicate))), mapped;
  }
  splice(index, deleteCount = 1) {
    const result = this.value.splice(index, deleteCount);
    return this.messager.dispatch(this.value), result;
  }
  *[Symbol.iterator]() {
    yield* this.value;
  }
}
((StateArray2) => {
  function from(init) {
    const state = new StateArray2(Signal.get(init) ?? []);
    return Signal.subscribe(init, (value) => state.set(value ?? [])), state;
  }
  StateArray2.from = from;
  function fromAsync(init) {
    const state = new StateArray2();
    return State.subscribeImmediate(init, (value) => {
      Promise.resolve(value).then((x) => state.set(x ?? []));
    }), state;
  }
  StateArray2.fromAsync = fromAsync;
})(StateArray || (StateArray = {}));
class StateBoolean extends State {
  static {
    this.TOGGLE_FUNCTION = (value) => !value;
  }
  toggle() {
    this.set(StateBoolean.TOGGLE_FUNCTION);
  }
}
class Emitter {
  constructor() {
    this.callbacks = {}, this.callbacksAny = /* @__PURE__ */ new Set();
  }
  on(event, callback) {
    this.callbacks[event] ??= /* @__PURE__ */ new Set(), this.callbacks[event]?.add(callback);
  }
  off(event, callback) {
    this.callbacks[event]?.delete(callback);
  }
  once(event, callback) {
    const once = (value) => {
      callback(value), this.off(event, once);
    };
    this.on(event, once);
  }
  until(event, timeout = 15 * 1e3) {
    return new Promise((resolve, reject) => {
      const timeoutId = setTimeout(() => reject(new Error(`Time's out awaiting event "${event.toString()}"`)), timeout);
      this.once(event, (value) => {
        resolve(value), clearTimeout(timeoutId);
      });
    });
  }
  dispatch(event, payload) {
    this.callbacks[event]?.forEach((callback) => callback(payload)), this.callbacksAny.forEach((callback) => callback(this));
  }
  when(event) {
    return {
      subscribe: (next) => {
        const callback = (value) => next?.(value);
        return this.on(event, callback), { unsubscribe: () => this.off(event, callback) };
      }
    };
  }
  subscribe(next) {
    return this.callbacksAny.add(next), { unsubscribe: () => void this.callbacksAny.delete(next) };
  }
  toJSON() {
  }
}
class EventSignal {
  constructor(initialValue) {
    this.messager = new Messager(), this.value = initialValue;
  }
  get current() {
    return this.value;
  }
  get() {
    return this.value;
  }
  set(newValue) {
    const value = newValue instanceof Function ? newValue(this.value) : newValue;
    this.value = value, this.messager.dispatch(value);
  }
  subscribe(next) {
    return this.messager.subscribe(next);
  }
  [Symbol.subscribe](next) {
    return this.messager.subscribe(next);
  }
  toJSON() {
    return this.value;
  }
  valueOf() {
    return this.value;
  }
}
class StateFSM extends EventSignal {
  constructor() {
    super(...arguments), this.events = new Emitter();
  }
  when(state) {
    return this.events.when(state);
  }
  set(newValue) {
    super.set(newValue), this.events.dispatch(this.value, void 0);
  }
}
export {
  ClosureCaptor,
  ClosureParticipant,
  ClosureSignal,
  Emitter,
  EventSignal,
  Messager,
  Notifier,
  State,
  StateArray,
  StateBoolean,
  StateFSM
};
